### PR TITLE
Refactor encode-decode tests to reduce duplication

### LINKED ISSUE #956: Reduce duplication in encode-decode tests
Right now everything is duplicated, which is very annoying for making changes.

We have two separate but nearly identical implementations, one using an internal copy of aom and one using [dav1d-sys](http://crates.io/crates/dav1d-sys):
```
src/test_encode_decode_aom.rs 
src/test_encode_decode_dav1d.rs
```
- [ ] Move the decoder-specific code in a trait
- [ ] Move all the shared code in `src/test_encode_decode.rs` and make it use the trait you created
- [ ] Use [interpolate_test](https://crates.io/crates/interpolate_name) and `cfg_attr` to not repeat yourself

### PR DESCRIPTION
**Pull request by shssoichiro**
Closes #956

Some dav1d tests are known to fail due to rounding errors.
This is an existing issue and is not resolved in this commit.

### CODE REVIEW

File: src/test_encode_decode.rs [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+    assert_eq!(&line.0[..width], &line.1[..width]);
+  }
+}
+
+pub(crate) fn setup_encoder<T: Pixel>(
+  w: usize, h: usize, speed: usize, quantizer: usize, bit_depth: usize,
+  chroma_sampling: ChromaSampling, min_keyint: u64, max_keyint: u64,
+  low_latency: bool, bitrate: i32
+) -> Context<T> {
+  unsafe {
```
Comment (lu-zero) [inline src/test_encode_decode.rs]: This should go away.

**Comment by lu-zero** [comment]
Has the problem already been reported to upstream dav1d btw?

**Comment by shssoichiro** [comment]
Does not appear so. I can file a report upstream.

File: src/test_encode_decode.rs (hunk 1 above)
Comment (shssoichiro) [reply src/test_encode_decode.rs]: Removed

**Comment by shssoichiro** [comment]
Actually, it's possible that this is a rounding error in rav1e. If I change the seed of the test RNG from `[0; 32]` to `[1; 32]`, one of the AOM tests also fails on the same data as the dav1d test. Although it's strange that that isn't happening with the existing RNG seed.

**Comment by shssoichiro** [comment]
I think this depends on https://github.com/xiph/rav1e/pull/1046 for the tests to go green.

**Comment by tdaede** [comment]
Or all of the above. Can you dump the [0; 32] seed IVF?

**Comment by shssoichiro** [comment]
Yes, here is the IVF from one of the failing tests (`bitrate_dav1d`): [file](condor3280.startdedicated.com/files/out-64x80-s10-q172-Cs420.ivf)

File: src/test_encode_decode_dav1d.rs [hunk 2]
```diff
@@ -8,272 +8,127 @@
...
-  }
-}
+          let packet = pkt.data;
 
-macro_rules! test_speeds {
-  ($($S:expr),+) => {
-    $(
-        paste::item!{
-            #[test]
-            #[ignore]
```
Comment (lu-zero) [inline src/test_encode_decode_dav1d.rs]: You should get the `#[ignore]` markers in the new code as well.

File: src/test_encode_decode_dav1d.rs [hunk 3]
```diff
@@ -8,272 +8,127 @@
...
-  let q = 100;
-
-  for keyint in &[4, 5, 6] {
-    encode_decode(w, h, speed, q, limit, 8, Default::default(), *keyint, *keyint, false, 0);
-  }
-}
-
-#[test]
-#[ignore]
-fn odd_size_frame_with_full_rdo() {
```
Comment (lu-zero) [inline src/test_encode_decode_dav1d.rs]: This seems to be the current offender, adding `#[ignore]` to this test should keep travis happy.

**Pull request merged**